Update ExternalMap and relationships to reference Artifacts (#542)
spdx/spdx-3-model: model/Core/Vocabularies/RelationshipType.md

- republishedBy: (Security) Designates the agent that tracked, aggregated, and/or enriched vulnerability details to improve context (i.e. NVD)
- requirementFor: Every `to` Element is required for the `from` Element
- runtimeDependency: Every `to` Element is a runtime dependency for the `from` Element

> - serializedArtifactFor: Every `to` Element is an artifact serialized from the `from` SpdxDocument Element

Reviewer: I would strongly recommend this RelationshipType be 'from' an "XCollection" 'to' an Artifact and not the other way around.
A given Artifact is never going to be a serialization of multiple XCollections but a given XCollection will often have multiple serializations represented by different Artifacts.

I also strongly suggest the RelationshipType string "hasSerialization" as it is extremely clear and descriptive in its meaning and unlikely to be misunderstood.

Reviewer: @goneall I think the modified description is correct but the relationshipType name is still confusing and is open to misinterpretation of directionality.
If you look at the set of other RelationshipType names I think you will find that the most clear and least confusing are the ones that follow a pattern of "<verb> <preposition>" or "<verb> <noun>". These have clarity of directionality, clarity of the nature of the relationship between the 'to' element and the 'from' element(s), and they can easily be read in the fashion "<'from' element> <RelationshipType> <'to' element(s)>" (e.g., "https://foo.org/spdx/element-4f888b42-12f0-4183-88b8-3b6d60162243" "dependsOn" "https://foo.org/spdx/element-55fc5492-b1b3-4152-a2ab-206e763e2bf7"). Current RelationshipType names that do not follow this pattern (e.g., "devTool", "example", "prerequisite", etc.) are far more ambiguous and open to confusion or misinterpretation. Simply adding a "has" to the beginnings of most of these (e.g., "hasDevTool", "hasExample", "hasPrerequisite") adds a great deal of clarity and significantly reduces the likelihood of confusion or misinterpretation.

Again, I would strongly recommend using "hasSerialization" as the RelationshipType name. The description can provide the detail of what types of elements are valid in the 'to' and 'from position (and this can be formally added to the SHACL once the spec generator is ready to tackle this). It is not needed in the name.

Reviewer: Oops. I just realized that the directionality in the description is reversed. I misread it.

It should be 'from' and SpdxDocument 'to' an Artifact. This is called out in the comment below and is always whatI supported (as my proposal diagrams all show).
I apologize for the confusion.

Reviewer 2: The name is still confusing IMHO. Relationships are most clear if they complete the sentence `from` (is) (a) `RELATIONSHIP` `to`. In this case, you are saying:
`SpdxDocument` is a `serializedArtifactFor` `Artifact`

The confusing part here is that the `spdxDocument` is not the serialized artifact (at least AFACIT, correct me if I'm wrong?)

More clear would be either to flip the `from` and `to`
`Artifact` is a `serializedArtifactFor` `SpdxDocument`

or reword:

`SpdxDocument` is `serializedFromArtifact` `Artifact`

Reviewer 2: Also, it's helpful if your documentation matches the `from` -> `to` in it's ordering also, so I would say:

`serializedFromArtifact`: The `from` SpdxDocument was serialized by the `to` Artifact

Author: @JPEWdev @sbarnum - if you could leave a very specific suggestion to the updated name and documentation, I'll accept it in the PR.  Otherwise, we may be going back and forth a bit before I completely meet your expected result ;)

> - spdxDocumentForArtfact: Every `to` Element is an SpdxDocument representing the logical serialization metadata for the `from`  Artifact containing the serliazed data

Author: ```suggestion
- serializedArtifactFor: Every `to` Element is an artifact serialized from the `from` SpdxDocument Element
```

@sbarnum - when I tried to implement the suggested direction in the serialization example, it didn't seem to work - you would want to relate one logical SPDX document to multiple artifacts - not the other way around.  Thinking of changing this back.  What do you think?

Reviewer: Oops, you are correct. I just realized that the directionality in the description is reversed. I misread it.

It should be 'from' and SpdxDocument 'to' an Artifact. 
This is what my initial comment above proposed the change be ("I would strongly recommend this RelationshipType be 'from' an "XCollection" 'to' an Artifact and not the other way around.").

I apologize for the confusion.
